src/Build.UnitTests/BuildEventArgsDataEnumeration.cs

```diff
@@ -0,0 +1,86 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Build.Execution;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Logging;
+using Microsoft.Build.Shared;
+using Shouldly;
+using Xunit;
+
+namespace Microsoft.Build.UnitTests
+{
+    public class BuildEventArgsDataEnumeration
+    {
+        [Fact]
+        public void SamplePropertiesEnumeration()
+        {
+            var projectFile = @"C:\foo\bar.proj";
+            var args = new ProjectEvaluationFinishedEventArgs(
+                ResourceUtilities.GetResourceString("EvaluationFinished"),
+                projectFile)
+            {
+                BuildEventContext = BuildEventContext.Invalid,
+                ProjectFile = @"C:\foo\bar.proj",
+                GlobalProperties = new Dictionary<string, string>() { { "GlobalKey", "GlobalValue" } },
+                Properties = new List<object>()
+                {
+                    new DictionaryEntry("Key", "Value"),
+                    ProjectPropertyInstance.Create("prop", "val"),
+                    new KeyValuePair<string, string>("foo","bar")
+                },
+                Items = null
+            };
+
+            List<(string propertyName, string propertyValue)> results = args.EnumerateProperties().ToList();
+            results.Count.ShouldBe(3);
+            results[0].ShouldBe(("Key", "Value"));
+            results[1].ShouldBe(("prop", "val"));
+            results[2].ShouldBe(("foo", "bar"));
+        }
+
+        [Fact]
+        public void SampleItemsEnumeration()
+        {
+            string projectFile = @"C:\foo\bar.proj";
+            ProjectEvaluationFinishedEventArgs args = new ProjectEvaluationFinishedEventArgs(
+                ResourceUtilities.GetResourceString("EvaluationFinished"),
+                projectFile)
+            {
+                BuildEventContext = BuildEventContext.Invalid,
+                ProjectFile = @"C:\foo\bar.proj",
+                GlobalProperties = new Dictionary<string, string>() { { "GlobalKey", "GlobalValue" } },
+                Properties = null,
+                Items = new List<DictionaryEntry>()
+                {
+                    new DictionaryEntry("Key", new MyTaskItem() { ItemSpec = "TestItemSpec" }),
+                    new DictionaryEntry("Key2",
+                        new TaskItemData("spec",
+                            new Dictionary<string, string>() { { "metadat1", "val1" }, { "metadat2", "val2" } })),
+                }
+            };
+
+            List<(string itemType, IItemData itemValue)> results = args.EnumerateItems().ToList();
```

Comment:
would it be better to create a data type instead of the types specified for the Tuple in the generic List?

Reply:
This is a fair concern. We cannot go wrong with explictly defined record here - I'll update.
Thanks!

Reply:
Explicit types added